Libplanet/Net/TxCompletion.cs

```diff
@@ -111,95 +99,123 @@ private HashSet<TxId> GetRequiredTxIds(IEnumerable<TxId> ids)
-                                    "since it does not follow policy.",
-                                    tx.Id);
-                                _blockChain.StagePolicy.Ignore(_blockChain, tx.Id);
-                                return false;
-                            }));
-
-                    var validTxs = new List<Transaction<TAction>>();
-                    IImmutableSet<TxId> stagedTxIds = _blockChain.GetStagedTransactionIds();
-                    foreach (var tx in txs)
-                    {
-                        try
+                _logger.Debug(
+                    "Start to run _txFetcher from {Peer}. (count: {Count})",
+                    peer,
+                    txIds.Count);
+                var stopWatch = new Stopwatch();
+                stopWatch.Start();
+                var txs = new HashSet<Transaction<TAction>>(
+                    await _txFetcher(
+                            peer,
+                            txIds,
+                            cancellationToken)
+                        .ToListAsync(cancellationToken)
+                        .AsTask());
+                _logger.Debug(
+                    "End of _txFetcher from {Peer}. (received: {Count}); " +
+                    "Time taken: {Elapsed}",
+                    peer,
+                    txs.Count,
+                    stopWatch.Elapsed);
+
+                txs = new HashSet<Transaction<TAction>>(
+                    txs.Where(
+                        tx =>
                         {
-                            if (!stagedTxIds.Contains(tx.Id))
+                            if (_blockChain.Policy.ValidateNextBlockTx(
+                                    _blockChain,
+                                    tx) is null)
                             {
-                                _blockChain.StageTransaction(tx);
-                                validTxs.Add(tx);
+                                return true;
                             }
-                        }
-                        catch (InvalidTxException ite)
-                        {
-                            _logger.Error(
-                                ite,
-                                "Transaction {TxId} will not be staged since it is invalid.",
+
+                            _logger.Debug(
+                                "Received transaction from {Peer} with id " +
+                                "{TxId} will not be staged " +
+                                "since it does not follow policy.",
+                                peer,
                                 tx.Id);
+                            _blockChain.StagePolicy.Ignore(_blockChain, tx.Id);
+                            return false;
+                        }));
+
+                var validTxs = new List<Transaction<TAction>>();
+                IImmutableSet<TxId> stagedTxIds = _blockChain.GetStagedTransactionIds();
```

Comment:
`BlockChain<T>.StageTransaction()` should do the job of filtering so no additional filtering is necessary at this point. See #1648.

Reply:
Eh, then again, we can't count `validTxs` if we don't do the filtering here. I'll create an issue for this.

Reply:
So you mean L158 to L176 is not required?

Reply:
I'll deal with this in #1705.

Reply:
`IStagePolicy` is designed not to throw any `Exception` so yes, I think it should be fine.

Reply:
But yeah, since broadcasting is tied to `validTxs`, I think it should be left as it is for now.